extra_data/keydata.py

```diff
@@ -216,33 +218,128 @@ def as_single_value(self, rtol=1e-5, atol=0.0, reduce_by='median'):
 
     # Getting data as different kinds of array: -------------------------------
 
-    def ndarray(self, roi=(), out=None):
+    def _read_tasks(self, parts=1):
+        """Produce file chunks with corresponding slices of the output array"""
+        dest_cursor = 0
+        for part in self.split_trains(parts=parts):
+            for chunk in part._data_chunks_nonempty:
+                dest_chunk_end = dest_cursor + chunk.total_count
+                yield chunk, np.s_[dest_cursor:dest_chunk_end]
+                dest_cursor = dest_chunk_end
+
+    # In worker processes, we need the shared output array accessible as global
+    # without copying it. Each object launches its own worker processes to
+    # read data, so global variables in the workers are safe.
+    _out_arr = None
+    @classmethod
+    def _set_out_arr(cls, out):
+        cls._out_arr = out
+
+    @classmethod
+    def _read_chunk(cls, chunk, dest_slice, roi):
+        out = cls._out_arr
+        chunk.dataset.read_direct(
+            out[dest_slice], source_sel=(chunk.slice,) + roi
+        )
+
+    @staticmethod
+    def _read_direct_frames(chunk):
+        return [
+            chunk.dataset.id.read_direct_chunk((i, 0, 0, 0))[1]
+            for i in range(chunk.first, int(chunk.first) + chunk.total_count)
+        ]
+
+    def _read_compressed_frames(self, out, roi=(), read_procs=1, decomp_threads=1):
+        """Read frame-wise compressed data (mask/gain) for 2D detectors,
+        decompressing data in parallel"""
+        from multiprocessing.pool import Pool, ThreadPool
+        import zlib
+
+        if read_procs < 0:
+            read_procs = 10  # Seems to be a decent guess on GPFS
+        if decomp_threads < 0:
+            decomp_threads = available_cpu_cores()
+
+        if read_procs == 1:
+            all_compressed_frames = sum([
+                self._read_direct_frames(chk) for chk in self._data_chunks_nonempty
+            ], [])
+        else:
+            chunks = sum([
+                p._data_chunks_nonempty for p in self.split_trains(parts=read_procs)
+            ], [])
+            with Pool(processes=read_procs) as ppool:
+                chks_compressed = ppool.map(self._read_direct_frames, chunks)
+                all_compressed_frames = sum(chks_compressed, [])
+
+        def decompress_frame(target_ix, compressed):
+            shuffled = zlib.decompress(compressed)
+            # Equivalent to the HDF5 'shuffle' filter: transpose bytes
+            data = np.ascontiguousarray(
+                np.frombuffer(shuffled, dtype=np.uint8).reshape((out.itemsize, -1)).transpose()
+            ).view(out.dtype).reshape(1, 512, 1024)
+            out[target_ix] = data[roi]
+
+        with ThreadPool(decomp_threads) as tpool:
+            tpool.starmap(decompress_frame, enumerate(all_compressed_frames))
+
+        return out
+
+    def ndarray(self, roi=(), out=None, read_procs=1, decomp_threads=1):
```

Comment:
`decomp_threads` failed for me when loading an AGIPD module with shape `(200000, 512, 128)`:
![image](https://user-images.githubusercontent.com/5361518/206248526-8c45b11c-7543-4c1b-908a-015a6a54bd92.png)